src/main/java/com/gluonhq/substrate/util/ios/MobileDeviceBridge.java

```diff
@@ -0,0 +1,449 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.substrate.util.ios;
+
+import com.dd.plist.NSArray;
+import com.dd.plist.NSDictionary;
+import com.dd.plist.NSObject;
+import com.dd.plist.PropertyListParser;
+import com.gluonhq.substrate.util.FileOps;
+import com.gluonhq.substrate.util.Logger;
+import com.gluonhq.substrate.util.ios.Device.InstproxyStatusCallback;
+import jnr.ffi.LibraryLoader;
+import jnr.ffi.Memory;
+import jnr.ffi.NativeType;
+import jnr.ffi.Pointer;
+import jnr.ffi.Runtime;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ *
+ * @author johan
+ */
+public class MobileDeviceBridge {
+
+    public static final MobileDeviceBridge instance = new MobileDeviceBridge();
+
+    private Device device = null;
+    private Runtime runtime;
+
+    public final static String AFC_SERVICE_NAME = "com.apple.afc";
+    public final static String INSTPROXY_SERVICE_NAME = "com.apple.mobile.installation_proxy";
+
+    private MobileDeviceBridge() {
+        Logger.logDebug("MobileDeviceBridge :: Create bridge");
+    }
+
+    public void init() {
+        Logger.logDebug("MobileDeviceBridge Prepare device");
+
+        Path lib = null;
+        try {
+            lib = FileOps.copyResourceToTmp("/thirdparty/libimobiledevice.dylib");
+        } catch (IOException e) {
+            Logger.logSevere(e, "Error copying libimobiledevice.dylib");
+        }
+        device = LibraryLoader.create(Device.class).load(lib.toString());
+        device.idevice_set_debug_level(10);
+        runtime = jnr.ffi.Runtime.getRuntime(device);
+    }
+
+    public boolean isReady() {
+        return device != null;
+    }
+
+    public String[] getDeviceIds() {
+        if (!isReady()) {
+            Logger.logSevere("Error: No device was found");
+            return null;
```

Comment:
I think it has to throw the custom exception

Reply:
I wouldn't throw an exception, it happens simply by not plugging your phone, and there are other checks later on.

Reply:
As long as the process is stopped after this..

Reply:
I would add an exception too. Otherwise, your calling code need to do null checks. This is probably only called for deploying code to the device, hence when there is no device, there is no point in continuing?

Reply:
Actually I've just realized that  `isReady()` checks for `device` (the interface), not for connected `devices`, so in case it is null, we need to call `init()`.

Reply:
"we" as in `getDeviceIds()` or the caller of this function?
Whatever the answer is, it needs to be added to the JavaDoc of this method, so that the caller knows what she can expect (e.g. "it's totally fine to get `null` and in that case I might decide to call `init()` " or "if this throws an exception, it's impossible to use the device now or later" or....)

Reply:
I moved `init` to private, and to be called from `getDevicesIds()`. So far, this is the only call we do to `MobileDeviceBridge`. With `ios-deploy` we might even remove its use at all.